feat(button): pending state

mdt2 opened the pull request:
## Description

Add pending state variant to the existing button

## How and where has this been tested?

Please tag yourself on the tests you've marked complete to confirm the tests have been run by someone other than the author.

### Validation steps
#### New/additional validation steps
- [ ] The "iconOnly" version of the button (aka the circular one) remains a perfect circle when the pending state is applied. Check that the height and width match and that the iconOnly pending state button has the same height as other variants of the button. This is best checked in the docs site but can be tested in Storybook as well.
- [ ] In Storybook, Pending has it's own story that shows pending state default and static white with labels (Cassondra recommended this and is implementing similar things in her [decorator PR](https://github.com/adobe/spectrum-css/pull/2420/files#diff-b965736461e30accd067d0f9fe09aa0e60ac58b4910b6dfb39c194bbb5200f5a))
- [ ] WithForcedColors story includes Pending state
- [ ] Test WHCM. Buttons should look like this per chats with the accessibility team:
![Screen Shot 2024-01-23 at 2 25 48 PM](https://github.com/adobe/spectrum-css/assets/23404383/737624c4-367b-41c6-a8c3-731fbc8737cc)


#### Original validation steps
- [x] Design approval @mdt2 
  - Outstanding design questions [and answers](https://adobedesign.slack.com/archives/G019JTYMT6H/p1705609053178079?thread_ts=1705006780.231489&cid=G019JTYMT6H):
    - [x] Should the pending button be disabled to prevent user interaction and hover/focus states? **yes**
    - [x] Assuming it is disabled, the background color and border should be the same no matter what background color variant the button originally was, correct? For example even if it is a negative variant button, when it is pending it should look like the pending spec with the disabled gray background/border and the default blue for the progress circle? **yes, it can use disabled-background-color and disabled-border-color**
    - [x] What size Progress Circle (PC) should I use for each size of the button? **small for all size buttons, including desktop and mobile**
    - [x] What should the pending state look like against static white or black backgrounds? Should we use a different version of the Progress Circle in either case? **Static white uses the static white progress circle. Design is continuing to not support static black for progress circle, which means buttons on static black aren't eligible for pending state.**

- [x] Pending state has storybook representation @jenndiaz 
- [x] Pending state has docs site representation @jenndiaz 
- [x] Pending state matches [the spec](https://spectrum.adobe.com/page/button/#Pending) @jenndiaz 
- [x] ProgressCircle has increased chromatic coverage for indeterminate and static white variants @jenndiaz 

### Regression testing

Validate:

1. The documentation pages for at least two other components are still loading, including:

- [ ] The pages render correctly, are accessible, and are responsive.

2. If components have been modified, VRTs have been run on this branch:

- [ ] VRTs have been run and looked at.
- [ ] Any VRT changes have been accepted (by reviewer and/or PR author), or there are no changes.


## To-do list

<!-- Put an "x" to indicate you've done each of the following. Add/remove additional tasks, as needed. -->

- [x] I have read the [contribution guidelines](/.github/CONTRIBUTING.md).
- [x] I have updated relevant storybook stories and templates.
- [x] I have tested these changes in Windows High Contrast mode.
- [ ] If my change impacts **other components**, I have tested to make sure they don't break.
- [ ] If my change impacts **documentation**, I have updated the documentation accordingly.
- [ ] ✨ This pull request is ready to merge. ✨

mdt2 commented on components/button/metadata/button-accent.yml:
```diff
@@ -258,3 +258,90 @@ examples:
           </button>
         </div>
       </div>
+  - id: button-pending
```
A bit torn on how to display this in the docs site. Should it be it's own page? If not, should I show the same pending state examples for all button types since it will look the same for all of them (except maybe static color, waiting to hear back from design on this)?

pfulton commented on components/button/index.css:
```diff
@@ -229,11 +229,16 @@ a.spectrum-Button {
   &:disabled,
-  &.is-disabled {
+  &.is-disabled,
+  &[pending] {
```
Interesting! I know that Spectrum Web Components is using a `pending` attribute there, but I don't think this is a valid HTML attribute for a `button`. Perhaps we could leave this selector, but also add an `.is-pending` selector, as well?

mdt2 replied:
Yeah that's correct, it's not an HTML attribute. I'll push that update!

pfulton reviewed (commented):
Just one thing here, @mdt2

castastrophe commented on components/button/stories/button.stories.js:
```diff
@@ -136,6 +149,23 @@ const CustomButton = ({
 				hideLabel: true,
 				iconName: iconName ?? "Edit",
 			})}
+			${when(isPendingStateSupported, () =>
```
Should this queue off the isPending variable? I'm not seeing that included here at all.

mdt2 replied:
These two examples always show the pending state, so I've done this to hide them on the Static Black story since that one doesn't support pending state. The isPending variable still works, it just shows what the other buttons look like when pending state is applied. I liked the visibility of this option for users

castastrophe reviewed (approved):
One question about the use of isPending in the logic for storybook.

jenndiaz reviewed (approved):
Looks good to me

castastrophe commented:
Looking through the new baselines, I’m realizing that pending is a state where the others depicted are design variants, I’m not sure this fits in this set long-term. cc 
@pfulton thoughts on the vrt organization here?

castastrophe commented:
@mdt2 can you verify the high contrast snapshots? Does that look right to you? I’m not seeing a status on it.

pfulton commented:
> Looking through the new baselines, I’m realizing that pending is a state where the others depicted are design variants, I’m not sure this fits in this set long-term. cc @pfulton thoughts on the vrt organization here?

I could go either way on it, but maybe it makes the most sense to have it as a separate story. @mdt2 wanna make that change?

mdt2 commented:
I've made some updates since the previous approvals that need re-reviewed (I've also added more specific validation steps for these changes in the description):

- High contrast mode has been adjusted and approved by the accessibility team
- Identified and fixed a bug (when pending state was active, the button was no longer a perfect circle) for the iconOnly button
- Pending state has it's own story in Storybook now
- Pending state is included in the existing With Forced Colors story

mdt2 commented on components/button/index.css:
```diff
@@ -89,6 +100,15 @@ governing permissions and limitations under the License.
   --spectrum-button-top-to-text: var(--spectrum-button-top-to-text-extra-large);
   --spectrum-button-bottom-to-text: var(--spectrum-button-bottom-to-text-extra-large);
 
+  .spectrum--medium &[pending],
+  .spectrum--medium &.is-pending {
+    --mod-button-edge-to-visual-only: calc(3px + var(--spectrum-button-edge-to-visual-only));
+  }
+
+  .spectrum--large &[pending],
+  .spectrum--large &.is-pending {
+    --mod-button-edge-to-visual-only: calc(4px + var(--spectrum-button-edge-to-visual-only));
+  }
```
I couldn't find a token value that properly reflected the distance from button edge to the progress circle since it has different dimensions than the icons, so this is pretty custom. Should these be custom tokens? Or maybe there is a token value somewhere that I missed (I was [looking here](https://github.com/adobe/spectrum-tokens/blob/29663e130dc9f221462de87ae7df4847e325cb4b/packages/tokens/src/layout.json#L596))?

jawinn commented on components/button/index.css:
```diff
@@ -229,18 +249,28 @@ a.spectrum-Button {
   .spectrum-Button {
     --highcontrast-button-content-color-disabled: GrayText;
     --highcontrast-button-border-color-disabled: GrayText;
+    --mod-progress-circle-track-border-color: CanvasText;
```
Should this double border around the progress circle be set to `ButtonText` to ensure contrast when on top of `ButtonFace`?

jawinn commented on components/button/metadata/button-pending.yml:
```diff
@@ -0,0 +1,328 @@
+SpectrumSiteSlug: https://spectrum.adobe.com/page/button/#Pending
+examples:
+  - id: button-pending
+    name: Pending for accent, negative, primary, and secondary buttons
```
Would it be worth showing examples for pending on buttons that are accent, negative, primary, or secondary? It looks like these are not any of those variants and are showing on a default button without any variant class (that is looking like 'spectrum-Button--outline'---is button even supposed to be used without any variant set?). 

In Storybook, I can see that with accent, it will have a grey disabled background that I don't see on this page. You could show both outline and accent, since there only appears to be two different visual styles?

jawinn reviewed (changes requested):
This is looking great! Nice work. I have a few questions around what we're showing in the docs and one of the colors used in forced-colors mode.

For buttons that have label text and an optional icon, should they still be allowed to exist in the markup, and then be set to `display: none` in the styles when the pending class is present? I was testing things out in Storybook and I think if there is a label and text in there now like on the Default accent button story, they will still appear:
![Screenshot 2024-01-25 at 4 57 03 PM](https://github.com/adobe/spectrum-css/assets/965114/124a6aa2-b59e-40cd-b835-f2b6fd1fa9f9)

mdt2 commented:
@jawinn thank you for catching these, great thoughts! I've just pushed up those changes.
